### PR TITLE
✨ 💥 Support `transferMultiple` in tx history and tx page

### PR DESCRIPTION
**Pull request by Aludirk**

### CODE REVIEW

File: util/txInfo/txInfo.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  abiDecoder: null,
+  web3: null,
+  handlers: {
+    ETH: [
+      To.getEthTxTo,
+      From.getTxFrom,
+      Value.getEthTxValue,
+      Receipt.getEthTxReceipt,
+    ],
+    Transer: [
```
Comment (williamchong007) [inline util/txInfo/txInfo.js]: `Transfer`

File: pages/in/tx/_id/index.vue [hunk 2]
```diff
@@ -309,11 +341,17 @@ export default {
...
-      ]);
+      const [fromData, toData] = isMultipleId
+        ? [
+          await apiGetUserMinById(this.fromId).catch(() => ({})),
+          null,
+        ]
+        : await Promise.all([
+          apiGetUserMinById(this.fromId).catch(() => ({})),
+          apiGetUserMinById(this.displayToId).catch(() => ({})),
+        ]);
```
Comment (nwingt) [inline pages/in/tx/_id/index.vue]: Can be simplify into:
```javascript
const [fromData, toData] = await Promise.all([
  apiGetUserMinById(this.fromId).catch(() => ({})),
  isMultipleId ? null : apiGetUserMinById(this.displayToId).catch(() => ({})),
]);
```

File: components/TransactionHistory.vue [hunk 3]
```diff
@@ -334,6 +356,14 @@ export default {
     isTokensale(tx) {
       return tx.to === LIKE_COIN_ICO_ADDRESS;
     },
+    isMultipleIn(tx) {
```
Comment (williamchong007) [inline components/TransactionHistory.vue]: Rename to `isInMultiple` less ambigious

File: components/TransactionHistory.vue [hunk 4]
```diff
@@ -334,6 +356,14 @@ export default {
...
+      return (Array.isArray(tx.to)
+        && Array.isArray(tx.value)
+        && tx.to.includes(this.address));
+    },
+    isMultipleOut(tx) {
```
Comment (williamchong007) [inline components/TransactionHistory.vue]: Rename to `isOutMultiple`, less ambigious

File: components/header/TransactionHeader.vue [hunk 5]
```diff
@@ -182,6 +182,12 @@ export default {
     dateCompleted() {
       return new Date(this.timestamp * 1000).toString();
     },
+    effectiveToId() {
+      if (Array.isArray(this.toId) && (this.toId.length === 1)) {
+        return this.toId[0];
```
Comment (williamchong007) [inline components/header/TransactionHeader.vue]: for `this.toId.length > 1` should not return array directly, if `this.toId.length > 1` is an error case, should instant return empty string

File: components/header/TransactionHeader.vue [hunk 6]
```diff
@@ -3,14 +3,14 @@
     <div class="lc-container-2">
       <div class="lc-container-3 lc-bg-gray-1 icon-wrapper">
         <nuxt-link
-          v-if="toId"
-          :to="{ name: 'id', params: { id: toId } }"
+          v-if="effectiveToId && !Array.isArray(effectiveToId)"
```
Comment (williamchong007) [inline components/header/TransactionHeader.vue]: Should not check `!Array.isArray(effectiveToId)`, see comment below

File: util/EthHelper.js [hunk 7]
```diff
@@ -257,64 +241,34 @@ class EthHelper {
...
-    const txTo = decoded.params.find(p => p.name === '_to').value;
-    let _to = web3Instance.utils.toChecksumAddress(txTo);
-    let _from = isDelegated ? decoded.params.find(p => p.name === '_from').value : t.from;
-    _from = web3Instance.utils.toChecksumAddress(_from);
-    let _value = decoded.params.find(p => p.name === '_value').value;
+    TxInfo.init(abiDecoder, web3Instance);
+    const type = TxInfo.getTxType(t);
+    const _to = TxInfo.getTxTo(type, t);
+    const _from = TxInfo.getTxFrom(type, t);
+    const _value = TxInfo.getTxValue(type, t);
```
Comment (williamchong007) [inline util/EthHelper.js]: Merge all 4 functions into `const { type, to, from, value } = getTxInfo`

File: util/EthHelper.js [hunk 8]
```diff
@@ -257,64 +241,34 @@ class EthHelper {
...
         _from,
         _to,
         _value,
-      };
+      });
     }
     const [r, block] = await Promise.all([
       web3Instance.eth.getTransactionReceipt(txHash),
       web3Instance.eth.getBlock(t.blockNumber),
     ]);
```
Comment (williamchong007) [inline util/EthHelper.js]: Should merge above async logic into txinfo

File: util/EthHelper.js [hunk 9]
```diff
@@ -257,64 +241,34 @@ class EthHelper {
...
+      });
     }
-    return {
-      _to,
-      _from,
-      _value,
-      timestamp: block ? block.timestamp : 0,
-    };
+    return filterMultipleTransferTo(filterAddress,
+      TxInfo.getTxReceipt(type, t, r, block, _to, _from, _value));
```
Comment (williamchong007) [inline util/EthHelper.js]: Should reduce `getTxReceipt` interface to `(web3, abidecoder, (type?), t)`

File: util/EthHelper.js [hunk 10]
```diff
@@ -257,64 +241,34 @@ class EthHelper {
...
-    const isLock = decoded.name === 'transferAndLock';
-    if (decoded.name !== 'transfer' && !isDelegated && !isLock) {
-      throw new Error('Not LikeCoin Store transaction');
-    }
-    const txTo = decoded.params.find(p => p.name === '_to').value;
-    let _to = web3Instance.utils.toChecksumAddress(txTo);
-    let _from = isDelegated ? decoded.params.find(p => p.name === '_from').value : t.from;
-    _from = web3Instance.utils.toChecksumAddress(_from);
-    let _value = decoded.params.find(p => p.name === '_value').value;
+    TxInfo.init(abiDecoder, web3Instance);
```
Comment (williamchong007) [inline util/EthHelper.js]: Should not have to init a non-class, pass `abiDecoder, web3Instance` into each fn as param

File: util/txInfo/txInfo.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+import * as Value from './value';
+import * as Receipt from './receipt';
+
+
+const state = {
+  abiDecoder: null,
+  web3: null,
+  handlers: {
+    ETH: [
+      To.getEthTxTo,
```
Comment (williamchong007) [inline util/txInfo/txInfo.js]: should use key-value i.e. `ETH.to: To.getEthTxTo`

File: util/txInfo/txInfo.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+
+/**
+ * Initialize the transaction info handler.
+ *
+ * @param {abi-decoder} abiDecoder The ABI decoder for Etherem transactions.
+ * @param {web3} web3 The Web3 instance.
+ */
+export function init(abiDecoder, web3) {
+  if (state.abiDecoder === null) {
```
Comment (williamchong007) [inline util/txInfo/txInfo.js]: should not have to init

**Pull request merged**